### PR TITLE
Add a Lp3943 driver

### PR DESCRIPTION
**Pull request by KasperJSdeVries**
## Description

- Driver that can set leds to one of the 4 modes
- Can also set 2 dim registers.

## Motivation and Context

A Chip I use in a project with a custom board

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: devices/Lp3943/Lp3943.cs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+			WriteToRegister(register, toWrite);
+		}
+
+		private void SetDimPercentage(DimRegister dimRegister, int percentage)
+		{
+			var toWrite = (byte)((float)percentage / 100 * 256);
+			var register = dimRegister switch
+			{
+				Device.Lp3943.DimRegister.Dim0 => Register.Pwm0, Device.Lp3943.DimRegister.Dim1 => Register.Pwm1,
```
Comment (Ellerbach) [inline devices/Lp3943/Lp3943.cs:86]: can you write them on 2 lines? will be easier to read

File: devices/Lp3943/Lp3943.cs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+			WriteToRegister(Register.Ls0, ls0);
+			WriteToRegister(Register.Ls1, ls1);
+			WriteToRegister(Register.Ls2, ls2);
+			WriteToRegister(Register.Ls3, ls3);
+		}
+
+		/// <summary>
+		/// Resets the Lp3943 chip
+		/// </summary>
+		public void Reset()
```
Comment (Ellerbach) [inline devices/Lp3943/Lp3943.cs:144]: Does it has an interest to be exposed? is there a good case to reset it while already initialized and running?

File: devices/Lp3943/Lp3943.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+			_ledStates[led] = ledState;
+			Update();
+		}
+
+		/// <summary>
+		/// Sets the LEDs in the array to the assigned mode
+		/// </summary>
+		/// <param name="leds">LEDs to assign</param>
+		/// <param name="ledState">state to give the LEDs</param>
+		public void SetLed(int[] leds, LedState ledState)
```
Comment (Ellerbach) [inline devices/Lp3943/Lp3943.cs:169]: minor: please add the exception description as well

File: devices/Lp3943/Lp3943.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+		{
+			if (leds is null)
+				throw new ArgumentNullException(nameof(leds));
+
+			foreach (var led in leds)
+			{
+				if (led is <= 15 and >= 0)
+				{
+					_ledStates[led] = ledState;
+				}
```
Comment (Ellerbach) [inline devices/Lp3943/Lp3943.cs:179]: else throws an exception?

File: devices/Lp3943/README.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+The LP3943 is an integrated device capable of independently driving 16 LEDs. It uses I2C.
+
+## Documentation
+
+- LM75 [datasheet](https://www.ti.com/lit/ds/symlink/lp3943.pdf)
+
+## Usage
+
+placeholder text
```
Comment (Ellerbach) [inline devices/Lp3943/README.md:11]: with the few feedback, you'll be able to fill this part now ;-)

File: devices/Lp3943/samples/Program.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Device.I2c;
+using System.Threading;
+using Iot.Device.Lp3943;
+using Iot.Device.Lp3943.Samples;
+
+I2cConnectionSettings settings = new(3, Lp3943.DefaultI2cAddress);
```
Comment (Ellerbach) [inline devices/Lp3943/samples/Program.cs:9]: please check other bindings with I2C samples and copy/paste the block for ESP32, it's a commented block for setting up the pins. Also add it to the read me (same, just copy/paste)

File: devices/Lp3943/samples/Program.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+		6,
+		7,
+		8
+	},
+	LedState.Dim0
+);
+ledDriver.SetLed(9, LedState.Dim1);
+ledDriver.SetLed(10, LedState.Dim1);
+ledDriver.SetLed(11, LedState.Dim1);
+
```
Comment (Ellerbach) [inline devices/Lp3943/samples/Program.cs:51]: maybe add a Debug.WriteLine with the expected behavior or a simple comment in the code of what is expected here

File: devices/Lp3943/Lp3943.cs [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+			Thread.Sleep(5);
+			_controller.Write(_pinReset, PinValue.High);
+		}
+
+		/// <summary>
+		/// Sets led to assigned mode
+		/// </summary>
+		/// <param name="led">led to assign</param>
+		/// <param name="ledState">state to give the led</param>
+		public void SetLed(int led, LedState ledState)
```
Comment (Ellerbach) [inline devices/Lp3943/Lp3943.cs:156]: minor: add the exception description here. Also for the bindings, we try not to put too much text. So if the exception description is nicely done here, you can just remove from the exception itself. It actually saves quite some space. I know it's not done like that for all bindings but we try for new ones.

File: devices/Lp3943/Lp3943.cs [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+		}
+
+		/// <summary>
+		/// Sets the LEDs in the array to the assigned mode
+		/// </summary>
+		/// <param name="leds">LEDs to assign</param>
+		/// <param name="ledState">state to give the LEDs</param>
+		public void SetLed(int[] leds, LedState ledState)
+		{
+			if (leds is null)
```
[the remark this answers is not in the document]
Comment (KasperJSdeVries) [reply devices/Lp3943/Lp3943.cs:171]: I don't think it is our responsibility to force the developer to write efficient code, If they want to reassign the same led multiple times in one call that is their problem.

File: devices/Lp3943/Lp3943.cs (hunk 4 above)
Comment (KasperJSdeVries) [reply devices/Lp3943/Lp3943.cs:179]: Leaving a warning would be best but an exception is fine as well I guess

File: devices/Lp3943/Lp3943.cs [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+	/// </summary>
+	public class Lp3943 : IDisposable
+	{
+		/// <summary>
+		/// Lp3943 I2C Address
+		/// </summary>
+		public const byte DefaultI2cAddress = 0x60;
+
+		private readonly int _pinReset;
+		private GpioController _controller;
```
Comment (Ellerbach) [inline devices/Lp3943/Lp3943.cs:22]: This one can be readonly as well?

File: devices/Lp3943/Lp3943.cs [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+			var ls3 = FillSelectorRegister(_ledStates[12], _ledStates[13], _ledStates[14], _ledStates[15]);
+
+			WriteToRegister(Register.Ls0, ls0);
+			WriteToRegister(Register.Ls1, ls1);
+			WriteToRegister(Register.Ls2, ls2);
+			WriteToRegister(Register.Ls3, ls3);
+		}
+
+		/// <summary>
+		/// Resets the Lp3943 chip
```
Comment (Ellerbach) [inline devices/Lp3943/Lp3943.cs:150]:
```suggestion
		/// Resets the Lp3943 chip. This has effect only if you have set a proper Reset Pin.
```

File: devices/Lp3943/Lp3943.cs [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+			return result;
+		}
+
+		/// <summary>
+		/// Sets the dimming on a register
+		/// </summary>
+		/// <param name="register">register to dim</param>
+		/// <param name="frequency">frequency in Hz</param>
+		/// <param name="dimPercentage">percentage</param>
+		public void DimRegister(DimRegister register, int frequency, int dimPercentage)
```
Comment (Ellerbach) [inline devices/Lp3943/Lp3943.cs:120]: minor: add the exception description as well in the intellisense comments

File: devices/Lp3943/README.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+- LM75 [datasheet](https://www.ti.com/lit/ds/symlink/lp3943.pdf)
+
+![Lp3943 eval board example back side](https://raw.githubusercontent.com/nanoframework/nanoFramework.IoT.Device/develop/devices/Lp3943/Lp3943_eval_back.png)
+![Lp3943 eval board example front side](https://raw.githubusercontent.com/nanoframework/nanoFramework.IoT.Device/develop/devices/Lp3943/Lp3943_eval_front.png)
+
+## Usage
+
+**Important**: make sure you properly setup the I2C pins especially for ESP32 before creating the `I2cDevice`, make sure you install the `nanoFramework.Hardware.ESP32 nuget`:
+
+```cs
```
Comment (Ellerbach) [inline devices/Lp3943/README.md:16]:
```suggestion
```csharp
```

File: devices/Lp3943/README.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+//////////////////////////////////////////////////////////////////////
+// when connecting to an ESP32 device, need to configure the I2C GPIOs
+// used for the bus
+Configuration.SetPinFunction(21, DeviceFunction.I2C1_DATA);
+Configuration.SetPinFunction(22, DeviceFunction.I2C1_CLOCK);
+```
+
+For other devices like STM32, please make sure you're using the preset pins for the I2C bus you want to use.
+
+```cs
```
Comment (Ellerbach) [inline devices/Lp3943/README.md:26]:
```suggestion
```csharp
```

**Review by Ellerbach — COMMENTED** [review]
Very minors comments then we'll be good to go!

File: devices/Lp3943/Lp3943.cs (hunk 10 above)
Comment (KasperJSdeVries) [reply devices/Lp3943/Lp3943.cs:22]: With the way the Dispose pattern is implemented currently, no.

Comment (Ellerbach) [reply devices/Lp3943/Lp3943.cs:22]: if you don't set it up to null in the Dispose, it's all ok. Did I miss something else?

File: devices/Lp3943/Lp3943.cs (hunk 4 above)
Comment (Ellerbach) [reply devices/Lp3943/Lp3943.cs:199]: thanks. And to gain space, you can remove the text from the exception.

File: devices/Lp3943/Properties/AssemblyInfo.cs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+[assembly: AssemblyTitle("Iot.Device.Lm75")]
```
Comment (Ellerbach) [inline devices/Lp3943/Properties/AssemblyInfo.cs:5]:
```suggestion
[assembly: AssemblyTitle("Iot.Device.Lp3943")]
```

**Review by Ellerbach — APPROVED** [review]
Very little details and then good to go! Thanks again!

File: devices/Lp3943/Lp3943.cs (hunk 10 above)
Comment (KasperJSdeVries) [reply devices/Lp3943/Lp3943.cs:22]: no, then Ill change it 👍

**Comment by KasperJSdeVries** [comment]
I am currently doing some tests and the dim registers aren't fully working, I'm looking into it.

**Comment by KasperJSdeVries** [comment]
> I am currently doing some tests and the dim registers aren't fully working, I'm looking into it.

This has been fixed now

File: devices/Lp3943/Lp3943.nuspec [hunk 16]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
+  <metadata>
+    <id>nanoFramework.Iot.Device.Lp3943</id>
+    <version>$version$</version>
+    <title>nanoFramework.Iot.Device.Lp3943</title>
+    <authors>nanoFramework project contributors</authors>
+    <owners>nanoFramework,dotnetfoundation</owners>
```
Comment (josesimoes) [inline devices/Lp3943/Lp3943.nuspec:8]: Deprecated, please remove.

```suggestion
```

**Review by josesimoes — CHANGES_REQUESTED** [review]
@KasperJSdeVries look good! Thanks for adding another binding.

Besides the fix required in the nuspec, for which I've made the respective suggestions, can you please switch from tabs to spaces. That's what's used throughout the entire project code base.

**Review by josesimoes — APPROVED** [review]
👌🏻

**Pull request merged**